File: Reference MATLAB Files/ResonanceTest.m
clear all
close all

%% Settings
name = 'wf9148x12KxXXXX';

start = 5e3; % Sweep starting frequency in Hz
stop = 20e3; % Sweep stop frequency in Hz
points = 150; % Number of points for the sweep

ampIn = 2; % input amplitude in V
ampRange = 3; % oscillo amplitude in V

Lrec = 10000; % Recording length
Fs = 1e6; % Sampling freq

target = 12e3; % target resonance frequency in Hz
tolerance = 3e3; % tolerance for the past fail in Hz

%% Getting the data
system('taskkill /im MultiChannel.exe'); % Close TiePie in case it was open

if verLessThan('matlab', '8')
    error('Matlab 8.0 (R2012b) or higher is required.');
end

% Open LibTiePie and display library info if not yet opened:
import LibTiePie.Const.*
import LibTiePie.Enum.*

if ~exist('LibTiePie', 'var')
    % Open LibTiePie:
    LibTiePie = LibTiePie.Library;
end

% Enable network search:       
LibTiePie.Network.AutoDetectEnabled = true;

% Search for devices:
LibTiePie.DeviceList.update();

% figure('Name','Live plot')

figure('Name','Live plot')
vamprms = [];
ave = []; % average signal value
DatStor = {};
for incre = 1:points
    freq = start + (stop - start)/(points-1)*(incre-1);
    vfreq(incre) = freq;
    
    % Try to open an oscilloscope with block measurement support and a generator in the same device:
    clear scp;
    clear gen;
    for k = 0 : LibTiePie.DeviceList.Count - 1
        item = LibTiePie.DeviceList.getItemByIndex(k);
        if item.canOpen(DEVICETYPE.OSCILLOSCOPE) && item.canOpen(DEVICETYPE.GENERATOR)
            scp = item.openOscilloscope();
            if ismember(MM.BLOCK, scp.MeasureModes)
                gen = item.openGenerator();
                break;
            else
                clear scp;
            end
        end
    end
    clear item



    if exist('scp', 'var') && exist('gen', 'var')
        % Oscilloscope settings:

        % Set measure mode:
        scp.MeasureMode = MM.BLOCK;

        % Set sample frequency:
        scp.SampleFrequency = Fs; % 1 MHz

        % Set record length:
        scp.RecordLength = Lrec; % 10000 Samples

        % Set pre sample ratio:
        scp.PreSampleRatio = 0; % 0 %

        % For all channels:
        for ch = scp.Channels
            % Enable channel to measure it:
            ch.Enabled = true;

            % Set range:
            ch.Range = ampRange; 

            % Set coupling:
            ch.Coupling = CK.DCV; % DC Volt

            % Release reference:
            clear ch;
        end

        % Set trigger timeout:
        scp.TriggerTimeOut = 1; % 1 s

        % Disable all channel trigger sources:
        for ch = scp.Channels
            ch.Trigger.Enabled = false;
            clear ch;
        end

        % Locate trigger input:
        triggerInput = scp.getTriggerInputById(TIID.GENERATOR_NEW_PERIOD); % or TIID.GENERATOR_START or TIID.GENERATOR_STOP

        if triggerInput == false
            clear triggerInput;
            clear scp;
            clear gen;
            error('Unknown trigger input!');
        end

        % Enable trigger input:
        triggerInput.Enabled = true;

        % Release reference to trigger input:
        clear triggerInput;

        % Generator settings:

        % Set signal type:
        gen.SignalType = ST.SINE;

        % Set frequency:
        gen.Frequency = freq; 

        % Set amplitude:
        gen.Amplitude = ampIn; % 2 V

        % Set offset:
        gen.Offset = 0; % 0 V

        % Enable output:
        gen.OutputOn = true;

        % Print oscilloscope info:
        %display(scp);

        % Print generator info:
        %display(gen);
        
        % recover some information from generator
        vGenFreq(incre) = gen.Frequency;

        % Start measurement:
        scp.start();

        % Start signal generation:
        gen.start();

        % Wait for measurement to complete:
        while ~scp.IsDataReady
            pause(10e-3) % 10 ms delay, to save CPU time.
        end

        % Stop generator:
        gen.stop();

        % Disable output:
        gen.OutputOn = false;

        % Get data:
        arData = scp.getData();
        ave = [ave; mean(arData)];
        arData = arData - mean(arData);

        % Get all channel data value ranges (which are compensated for probe gain/offset):
        clear darRangeMin;
        clear darRangeMax;
        for i = 1 : length(scp.Channels)
            [darRangeMin(i), darRangeMax(i)] = scp.Channels(i).getDataValueRange();
        end

        % Plot results:
%         if mod(incre,40) == 1
%             figure
%             plot((1:scp.RecordLength) / scp.SampleFrequency, arData);
%             xlabel('Time [s]');
%             ylabel('Amplitude [V]');
%             legend('Channel 1','Channel 2');
%             grid on;
%         end

        % Close oscilloscope:
        clear scp;

        % Close generator:
        clear gen;
    else
        error('No oscilloscope available with block measurement support or generator available in the same unit!');
    end

    DatStor{incre} = arData;
    
    % Plot
    amprms = rms(arData-mean(arData));
    vamprms = [vamprms; amprms];
    
    plot(vfreq/1e3,vamprms*1e3)
    grid on;
    xlabel('Frequency [kHz]');
    ylabel('Oscillation amplitude [mV]');
    legend('Channel 1','Channel 2');
    
%     plot(Vvolt,rms(arData-mean(arData)));
%     grid on;
%     xlabel('Input piezo voltage (to be converted in length) [V]');
%     ylabel('Output voltage [mV]');
%     legend('Channel 1','Channel 2')
end

clear LibTiePie % necessary to use TiePie after MatLab

system('cd C:\Users\cri\Documents\Etienne & start MatLab.tps'); % Reopen TiePie

%% Post process 

% % Amplitude
% vamp = [];
% for k = 1:length(DatStor)
%     amp = max(DatStor{1,k})-min(DatStor{1,k});
%     vamp = [vamp; amp];
% end
% 
% % Plot
% figure('Name','min max')
% plot(vfreq,vamp,'-o')
% grid on;
% xlabel('Frequency [Hz]');
% ylabel('Amplitude [V]');
% legend('Channel 1','Channel 2');

% FFT
vfour = [];
for k = 1:length(DatStor)
    four = abs(fft(DatStor{1,k}-mean(DatStor{1,k})));
    four = four(1:Lrec/2+1,:);
    vfour = [vfour four];
end

% Plot
% figure('Name','fft')
% plot(Fs*(0:Lrec/2)/Lrec*1e-3,vfour);
% grid on;
% xlabel('Frequency [kHz]');
% ylabel('Amplitude');


% RMS
vamprms = [];
for k = 1:length(DatStor)
    amprms = rms(DatStor{1,k}-mean(DatStor{1,k}));
    vamprms = [vamprms; amprms];
end

[M,I] = max(vamprms(:,1));
normal = vamprms(:,1)/M;

cutofff = [interp1(normal(1:I),vfreq(1:I),0.5) interp1(normal(I:end),vfreq(I:end),0.5)];
Qfactor = vfreq(I)/(cutofff(2)-cutofff(1));
DampingRatio = 1/2/Qfactor;

% Plot
figure('Name','RMS')
subplot(6,3,[1:15])
plot(vfreq/1e3,vamprms*1e3)
grid on; hold on;
xlabel('Frequency [kHz]');
ylabel('Oscillation amplitude [mV]');
scatter(cutofff/1e3,[0.5 0.5]*M*1e3,'rx');
legend('Channel 1','Channel 2','Bandwidth frequency');

subplot(6,3,16); axis off;
text(0,0,['Resonance frequency :',newline,num2str(vfreq(I)/1e3),' kHz'])
subplot(6,3,17); axis off;
text(0,0,['Q factor : ',num2str(Qfactor)])
subplot(6,3,18); axis off;
text(0,0,['Damping ratio : ',num2str(DampingRatio)])

saveas(gcf,[name,' - RMS - ',datestr(now, 'dd-mmm-yyyy HH-MM-SS'),'.png']);

% % Plot
% figure('Name','Average signal')
% plot(vfreq/1e3,ave*1e3)
% grid on;
% xlabel('Frequency [kHz]');
% ylabel('Amplitude [mV]');
% legend('Channel 1','Channel 2');
% 
% mean(ave)
% 
% % % Plot
% normal = vamprms(:,1)./vamprms(:,2);
% % figure('Name','RMS normalised')
% % plot(vfreq/1e3,normal)
% % grid on;
% % xlabel('Frequency [kHz]');
% % ylabel('Amplification factor');

%% Peaks
figure('Name','RMS peaks')
findpeaks(normal,vfreq/1e3,'MinPeakHeight',2*mean(normal))
grid on;
xlabel('Frequency [kHz]');
ylabel('Amplification factor');

[pks,loc] = findpeaks(normal,vfreq/1e3,'MinPeakHeight',2*mean(normal));
cond = sum(loc >= (target-tolerance)/1e3 & loc <= (target+tolerance)/1e3);
if cond == 0
    disp('Failed chip')
else
    disp('Passed chip')
end


%% test
vamp = [];
for k = 1:length(DatStor)
    amp = mean(DatStor{1,k});
    vamp = [vamp; amp];
end

% % Plot
% figure('Name','amp mean')
% plot(vfreq/1e3,vamp-mean(vamp))
% grid on;
% xlabel('Frequency [kHz]');
% ylabel('Amplitude [V]');
% legend('Channel 1','Channel 2');
% 
% % Plot
% figure('Name','amp mean 2')
% plot(vfreq/1e3,vamp)
% grid on;
% xlabel('Frequency [kHz]');
% ylabel('Amplitude [V]');
% legend('Channel 1','Channel 2');

% % RMS
% vamprms = [];
% for k = 1:length(DatStor)
%     amprms = rms(DatStor{1,k}(Lrec/2:end,:)-mean(DatStor{1,k}(Lrec/2:end,:)));
%     vamprms = [vamprms; amprms];
% end

% % Plot
% figure('Name','RMS')
% plot(vfreq/1e3,vamprms)
% grid on;
% xlabel('Frequency [kHz]');
% ylabel('Amplitude [V]');
% legend('Channel 1','Channel 2');
% 
% % FFT
% vfour = [];
% for k = 1:length(DatStor)
%     four = abs(fft(DatStor{1,k}(Lrec/2:end,:)-mean(DatStor{1,k}(Lrec/2:end,:))));
%     four = four(1:Lrec/4+1,:);
%     vfour = [vfour four];
% end
% 
% % Plot
% figure('Name','fft')
% plot(Fs*(0:Lrec/4)/Lrec*2,vfour);
% grid on;
% xlabel('Frequency [Hz]');
% ylabel('Amplitude');

save([name,' - Resonance test - ',datestr(now, 'dd-mmm-yyyy HH-MM-SS')]);

%% Finder 
% [f,dummy] = ginput(1);
% point = round((f*1e3 - start)*(points-1)/(stop-start) + 1);
% 
% % Plot
% figure
% plot((0:Lrec-1)/Fs, DatStor{1,point});
% xlabel('Time [s]');
% ylabel('Amplitude [V]');
% legend('Channel 1','Channel 2');
% grid on;

%% Test 07/05/2021
% pos = 90;
% 
% figure
% plot((1:Lrec) / Fs, DatStor{1,pos}-mean(DatStor{1,pos}));
% xlabel('Time [s]');
% ylabel('Amplitude [V]');
% legend('Channel 1','Channel 2');
% grid on;

File: python-libtiepie-master/Python Routines/ResonanceTest.py
# -*- coding: utf-8 -*-
"""
Created on Wednesday 20.11.2021 at 10:44

Miraex - Resonance Test for piezoelectric shaker

@author: Claudio Jaramillo
"""

from __future__ import print_function
import libtiepie
import time
import matplotlib.pyplot as plt

import numpy as np
import MiraexLib.plot as miraex_plt
import MiraexLib.misc as miraex_misc
import MiraexLib.analysis as miraex_anls
from MiraexLib.printinfo import*


# %% Settings
name = input('Name for the file : ')

# Parameters for the sweep over frequency
start = 5e3  # Sweep starting frequency in Hz
stop = 20e3  # Sweep stop frequency in Hz
points = 150  # Number of points for the sweep

ampIn = 2  # input amplitude in V
ampRange = 3  # oscillo amplitude in V

Lrec = 10000  # Recording length for a data point
Fs = 1e6  # Sampling freq

target = 12e3  # target resonance frequency in Hz
tolerance = 3e3  # tolerance for the past fail in Hz

# %% Setting up result file
# Create file_name
name += 'ResonanceTest'
file_name = miraex_misc.CreateFileName(name, 'txt')
writeDir_func = miraex_misc.CreateWriteDir('raw')

# Open the text file to write the results
result_file = open(writeDir_func+file_name, "ab")
print(result_file)

result_file.truncate(0)

nb_of_points_to_show = int(input(
    'For how many iterations would you like to show the dynamic plot ? '))

# %% Opening devices and setting up devices

# Print library info:
print_library_info()

# Enable network search:
libtiepie.network.auto_detect_enabled = True

# Search for devices:
libtiepie.device_list.update()

# Parameters and lists that will be modified during the loop
# voltage amplitude RMS
v_amplitude_rms = []

# average signal value
average_signal = []

# Storage list
Data_Storage = []

# voltage at frequency
v_freq = []

# generator frequency
v_gen_freq = []

# Data with a removed DC component
data_no_offset = []

# Sweep loop
for incre in range(points):
    freq = start + (stop - start)/(points-1)*(incre-1)
    v_freq.append(freq)
    print('--------------')
    print('Current frequency : ', freq)

    # Try to open an oscilloscope with block measurement support and a generator in the same device:
    scp = None
    gen = None
    for item in libtiepie.device_list:
        if (item.can_open(libtiepie.DEVICETYPE_OSCILLOSCOPE)) and (item.can_open(libtiepie.DEVICETYPE_GENERATOR)):
            scp = item.open_oscilloscope()
            if scp.measure_modes & libtiepie.MM_BLOCK:
                gen = item.open_generator()
                break
            else:
                scp = None

    # Oscilloscope settings:
    # Set measure mode:
    scp.measure_mode = libtiepie.MM_BLOCK

    # Set sample frequency:
    scp.sample_frequency = Fs  # 1 MHz

    # Set record length:
    scp.record_length = Lrec  # 10000 samples

    # Set pre sample ratio:
    scp.pre_sample_ratio = 0  # 0 %

    # For all channels:
    for ch in scp.channels:
        # Enable channel to measure it:
        ch.enabled = True

        # Set range:
        ch.range = ampRange  # 8 V

        # Set coupling:
        ch.coupling = libtiepie.CK_DCV  # DC Volt

    # Set trigger timeout:
    scp.trigger_time_out = 1  # 1 s

    # Disable all channel trigger sources:
    for ch in scp.channels:
        ch.trigger.enabled = False

    # Locate trigger input:
    # or TIID_GENERATOR_START or TIID_GENERATOR_STOP
    trigger_input = scp.trigger_inputs.get_by_id(
        libtiepie.TIID_GENERATOR_NEW_PERIOD)

    if trigger_input is None:
        raise Exception('Unknown trigger input!')

    # Enable trigger input:
    trigger_input.enabled = True

    # Generator settings:

    # Set signal type:
    gen.signal_type = libtiepie.ST_SINE

    # Set frequency:
    gen.frequency = freq  # in Hz

    # Set amplitude:
    gen.amplitude = ampIn  # in V

    # Set offset:
    gen.offset = 0  # 0 V

    # Enable output:
    gen.output_on = True

    # Print oscilloscope info:
    # print_device_info(scp)

    # Print generator info:
    # print_device_info(gen)

    # Get information about the frequency
    v_gen_freq.append(gen.frequency)

    # Start measurement:
    scp.start()

    # Start signal generation:
    gen.start()

    # Wait for measurement to complete:
    while not scp.is_data_ready:
        time.sleep(0.01)  # 10 ms delay, to save CPU time

    # Stop generator:
    gen.stop()

    # Disable output:
    gen.output_on = False

    # Get data:
    data = scp.get_data()

    # Data processing

    # print(f'{data = }')
    # np_data will be rewritten at every loop iteration
    # np_data[nb of channels of the scope][nb of samples in channel]

    np_data = np.array(data)  # numpy array
    #print(f'{np_data.shape = }')

    # average value of the data fetched during 1 loop per channel
    # The first index is the iteration index
    # The second index is the avg for channel index
    average_signal.append([np.mean(np_data[i])
                           for i in range(len(np_data))])

    # Substract the mean of np_data to each element of np_data during 1 loop
    '''
    Processed data will have n number of elements, each has nb_of_channels sub-dimensions that represent the nb of channels that we are working with in the oscilloscope.

    The number of elements n is the number of iterations of the loop. It increases as the test time increases.
    '''
    # By substracting the mean we remove the DC offset for each channel
    data_no_offset.append(
        [np_data[i] - average_signal[incre-1][i] for i in range(len(np_data))])

    '''
    print('data_no_offset')
    print(data_no_offset)
    print('processed data shape = ', np.array(data_no_offset).shape)
    '''

    # Get all channel data value ranges (which are compensated for probe gain/offset)
    data_range_min_max = []
    for channel in scp.channels:
        data_range_min_max.append(
            [channel._get_data_value_min(), channel._get_data_value_max()])

    '''
    # Output CSV data:
    csv_file = open('OscilloscopeGeneratorTrigger.csv', 'w')
    try:
        csv_file.write('Sample')
        for i in range(len(data)):
            csv_file.write(';Ch' + str(i + 1))
        csv_file.write(os.linesep)
        for i in range(len(data[0])):
            csv_file.write(str(i))
            for j in range(len(data)):
                csv_file.write(';' + str(data[j][i]))
            csv_file.write(os.linesep)

        print()
        print('Data written to: ' + csv_file.name)

    finally:
        csv_file.close()
    '''

    # Close oscilloscope:
    del scp

    # Close generator:
    del gen

    # ---------------------------------------------------------------------------------------------
    # DATA POST PROCESSING
    # ---------------------------------------------------------------------------------------------

    Data_Storage.append(data)

    # RMS of the data fetched in 1 loop iteration, for each channel
    Amplitude_RMS = [np.sqrt(np.mean((np.array(data_no_offset[incre-1][i]))**2))
                     for i in range(len(data_no_offset[incre-1]))]

    v_amplitude_rms.append(Amplitude_RMS)
    v_amplitude_rms_np = np.array(v_amplitude_rms)

    # Create a Channel first array of RMS values for plotting and better addressing of the data
    channel_amp_rms = np.transpose(v_amplitude_rms_np)

    '''
    print('Amplitude RMS')
    print(Amplitude_RMS)

    print('v_amplitude_rms')
    print(v_amplitude_rms)
    '''

    x_data = v_gen_freq

    # y_data contains the RMS for iteration for each channel : y_data[iteration][channel]
    # In this case channel 0 is the measurement of amplitude of oscillation
    y_data = [v_amplitude_rms[i][0]*1000 for i in range(len(v_amplitude_rms))]
    print('Current RMS [mV] : ', v_amplitude_rms[incre][0]*1000)

    '''
    print('xdata')
    print(x_data)

    print('ydata')
    print(y_data)
    '''

    if incre < nb_of_points_to_show:
        # Dynamic plotting
        for i in range(len(channel_amp_rms)):
            miraex_plt.DynamicPlot2(v_gen_freq, (channel_amp_rms[i])*1000,
                                    'Frequency [Hz]', 'Oscillation Amplitude RMS [mV]', 'Dynamic Resonance Test', True)

    else:
        miraex_plt.CloseAllPlots()

# In the resulting file each column represents one channel
np.savetxt(result_file, v_amplitude_rms_np)
result_file.close()

time.sleep(1)

# Opening data file to plot final data
result_file = open(writeDir_func+file_name, "ab")
final_data = np.loadtxt(writeDir_func+file_name)

# Creating final data
final_data_tp = np.transpose(final_data)

# Plotting final data
x_data = np.array(v_gen_freq)
for i in range(len(final_data_tp)):

    # Create y data to plot per channel
    y_data = final_data_tp[i]

    # Find data maximum and minimum and index
    (my_max, my_max_index) = miraex_anls.find_max_peak(y_data)
    (my_min, my_min_index) = miraex_anls.find_min_peak(y_data)

    # Find and print the frequency at which the maximum appears
    resonance_frequency = x_data[my_max_index]
    print('Resonance Frequency : ', resonance_frequency)

    # We arbitrarily detect a noise only signal
    # We have to do this in order to avoind the FWHM detecting function breaking when trying to compute the FWHM for a noise only signal.
    # This is kind of a hack
    if my_max-my_min < 1e-3:
        continue

    # find the two crossing points
    hmx = miraex_anls.half_max_x(x_data, y_data)

    # print the FWHM
    fwhm = hmx[1] - hmx[0]
    print("FWHM:{:.3f}".format(fwhm))

    # Find the Q-Factor
    q_factor = resonance_frequency/fwhm
    print('Q Factor : ', q_factor)

    # a convincing plot
    half = max(y_data)/2.0

    legend = 'Channel '+str(i+1)
    caption = f'Resonance frequency : {resonance_frequency/1000:.2f} kHz | Q Factor : {q_factor:.2f}'

    miraex_plt.GenericPlot(
        x_data, y_data, 'Frequency', 'RMS', name, legend, x_log=True, is_caption=True, my_caption=caption)
    plt.plot(hmx, [half, half])


# Keep the ShowPlots command at the end of the script !!!!!!
miraex_plt.ShowPlots()
result_file.close()
